File: stat-scripts/dropin-stats.js
// #####################
// ##-  DropIn Statistics  -##
// ####################
'use strict';

// Required tables: Role, User, Users_Bar_Algo, Stats_DropIn

// Requires
var Parse = require('parse').Parse;
var _ = require('underscore');
var moment = require('moment');

// Parse Keys
Parse.initialize(process.env.PARSE_ID, process.env.PARSE_SECRET);

// Create new objects
var Role = Parse.Object.extend('Role');
var StatsDropIn = Parse.Object.extend('Stats_DropIn');
var UsersRewards = Parse.Object.extend('Users_Rewards');
var Timeline = Parse.Object.extend('Users_Timeline');

// Instantiate queries
var roleQuery = new Parse.Query(Role);
var usersQuery = new Parse.Query(Parse.User);
var usersRewardsQuery = new Parse.Query(UsersRewards);
var timelineQuery = new Parse.Query(Timeline);

// Main iteration
calcStats();

// Main function
function calcStats() {
  // Initialize master object to carry state of calculations through promises
  var data = {};

  // Stat Calculations
  roleQuery.equalTo('name', 'User');
  roleQuery.limit(1000);
  roleQuery.first().then(function(role) {
    return role;
  })
  .then(function(role) {
    usersQuery.equalTo('roleId', role);
    usersQuery.include('loyaltyLevelId');
    usersQuery.limit(1000);
    return usersQuery.find().then(function(userObjs) {
      // Initialize variables for stats
      var totalMales = 0;
      var totalFemales = 0;
      var totalGuests = 0;
      var totalRegulars = 0;
      var totalVips = 0;
      var age35Plus = 0;
      var age3034 = 0;
      var age2529 = 0;
      var age2124 = 0;

      // Total number of Drop In users
      var totalUsers = userObjs.length;

      _.each(userObjs, function(user) {
        // Number of males and females
        var gender = user.attributes.gender.toLowerCase();

        if (gender === 'male') {
          totalMales++;
        } else if (gender === 'female') {
          totalFemales++;
        }

        // Number of users by loyalty level
        var loyaltyLevel = user.attributes.loyaltyLevelId.attributes.name.toLowerCase();

        if (loyaltyLevel === 'guest') {
          totalGuests++;
        } else if (loyaltyLevel === 'regular') {
          totalRegulars++;
        } else if (loyaltyLevel === 'vip') {
          totalVips++;
        }

        // Number of users by age
        var dateGt35 = moment(new Date()).subtract(35, 'years');
        var dateGt30 = moment(new Date()).subtract(30, 'years');
        var dateGt25 = moment(new Date()).subtract(25, 'years');
        var dateGt21 = moment(new Date()).subtract(21, 'years');

        var dob = moment(user.attributes.dob);

        if (dob.isBefore(dateGt35)) {
          age35Plus++;
        } else if (dob.isBefore(dateGt30)) {
          age3034++;
        } else if (dob.isBefore(dateGt25)) {
          age2529++;
        } else if (dob.isBefore(dateGt21)) {
          age2124++;
        }
      });

      var stats = {
        calcDate: new Date(), // Point in time date tracking
        totalUsers: totalUsers,
        totalMales: totalMales,
        totalFemales: totalFemales,
        totalGuests: totalGuests,
        totalRegulars: totalRegulars,
        totalVips: totalVips,
        age35Plus: age35Plus,
        age3034: age3034,
        age2529: age2529,
        age2124: age2124
      };

      data.stats = stats;
      return userObjs;
    })
    .then(function(userObjs) {
      // Initialize more variables on data.stats object
      data.stats.totalActiveUsersByCredit = 0;
      data.stats.totalTrafficByCredit = 0;

      var date30DaysAgo = moment(new Date()).subtract(2, 'days');
      var startDay = moment().subtract(1, 'days').hours(9).minute(0).second(0).millisecond(0); // 9am yesterday UTC (4am EST)
      var endDay = moment().hours(9).minute(0).second(0).millisecond(0); // 9am today UTC (4am EST)

      var promise = Parse.Promise.as();
      _.each(userObjs, function(user) {
        var activeUserCreditCount = 0;
        var trafficCreditCount = 0;

        promise = promise.then(function() {
          timelineQuery.equalTo('userId', user);
          timelineQuery.limit(1000);
          return timelineQuery.find().then(function(results) {
            var filterByEventType = _.filter(results, function(obj) {
              if (obj.attributes.eventType === 'Credit Earned') {
                return obj;
              }
            });

            _.each(filterByEventType, function(result) {
              var lastCreditDate = moment(result.attributes.date);

              // Increase the active user count for any user who has visited any bar within the last 30 days
              if (result.attributes.date !== undefined && lastCreditDate.isAfter(date30DaysAgo)) {
                activeUserCreditCount++;
              }

              // Increase the traffic count for any user who visited any bar on today's date
              if (result.attributes.date !== undefined && lastCreditDate.isBetween(startDay, endDay)) {
                trafficCreditCount++;
              }
            });

            // Increment counts on data.stats object to be saved to the DB
            if (activeUserCreditCount > 0) {
              data.stats.totalActiveUsersByCredit++;
            }

            if (trafficCreditCount > 0) {
              data.stats.totalTrafficByCredit++;
            }
          });
        });
      });

      return promise;
    })
    .then(function() {
      var startDay = moment().subtract(1, 'days').hours(9).minute(0).second(0).millisecond(0); // 9am yesterday UTC (4am EST)
      var endDay = moment().hours(9).minute(0).second(0).millisecond(0); // 9am today UTC (4am EST)

      // Number of redeemed rewards
      data.stats.totalRewardsRedeemed = 0;

      var promise = Parse.Promise.as();
      usersRewardsQuery.limit(1000);
      return usersRewardsQuery.find().then(function(results) {
        _.each(results, function(obj) {
          promise = promise.then(function() {
            var userHasRedeemed = obj.attributes.userHasRedeemed;
            var redeemedOnDate = moment(obj.attributes.redeemedOnDate);

            // If a user has redeemed a reward, increase the count of the data.stats object to be saved to the DB
            if (userHasRedeemed && redeemedOnDate.isBetween(startDay, endDay)) {
              data.stats.totalRewardsRedeemed++;
            }
          });
        });

        return promise;
      });
    })
    .then(function() {
      var newStat = new StatsDropIn();

      newStat.save(data.stats).then(function(savedObj) {
        // success
        console.log('Parse record with object ID: ' + savedObj.id + ' has been successfully created.');
      }, function(error) {
        // error
        console.log('An error has occured: ' + error);
      });
    });
  });
}

File: stats-scripts/dropin-stats.rb
require 'active_record'
require 'parse-ruby-client'
require 'active_support/core_ext/date_time/calculations'
require 'active_support/core_ext/date/calculations'
require 'active_support/core_ext/time/calculations'
require 'mysql2'
require '../environments'

ActiveRecord::Base.establish_connection(
  adapter: 'mysql2',
  host: DB_HOST,
  database: DB,
  username: DB_USERNAME,
  password: DB_PASSWORD
)

parse = Parse.init application_id: PARSE_ID, api_key: PARSE_REST_API

class User < ActiveRecord::Base
end

class Timeline < ActiveRecord::Base
end

class RewardsUsers < ActiveRecord::Base
end

def user_role_object_id
  user_role_query = Parse::Query.new("Role")
  user_role_query.eq("name", "User")
  user_role = user_role_query.get.first["objectId"]
end

def guest_level_object_id
  guest_level_query = Parse::Query.new("Loyalty_Levels")
  guest_level_query.eq("name", "Guest")
  guest_level = guest_level_query.get.first["objectId"]
end

def regular_level_object_id
  regular_level_query = Parse::Query.new("Loyalty_Levels")
  regular_level_query.eq("name", "Regular")
  regular_level = regular_level_query.get.first["objectId"]
end

def vip_level_object_id
  vip_level_query = Parse::Query.new("Loyalty_Levels")
  vip_level_query.eq("name", "VIP")
  vip_level = vip_level_query.get.first["objectId"]
end

def save_stats(
  total_users,
  total_males,
  total_females,
  total_gender_unknown,
  total_guests,
  total_regulars,
  total_vips,
  age_35plus,
  age_3034,
  age_2529,
  age_2124,
  active_users,
  traffic,
  rewards_redeemed
)
  dropin_stats = Parse::Object.new("Stats_DropIn")

  dropin_stats["calcDate"] = Parse::Date.new(Date.yesterday)
  dropin_stats["totalUsers"] = total_users
  dropin_stats["totalActiveUsersByCredit"] = active_users
  dropin_stats["totalTrafficByCredit"] = traffic
  dropin_stats["totalRewardsRedeemed"] = rewards_redeemed
  dropin_stats["totalMales"] = total_males
  dropin_stats["totalFemales"] = total_females
  dropin_stats["totalGenderUnknown"] = total_gender_unknown
  dropin_stats["totalGuests"] = total_guests
  dropin_stats["totalRegulars"] = total_regulars
  dropin_stats["totalVips"] = total_vips
  dropin_stats["age2124"] = age_2124
  dropin_stats["age2529"] = age_2529
  dropin_stats["age3034"] = age_3034
  dropin_stats["age35Plus"] = age_35plus

  dropin_stats.save

  puts "Drop In stats record successfully created"
end

def calc_stats()
  start_calc_datetime = 1.day.ago.change({ hour: 9, min: 0, sec: 0, usec: 0 }).iso8601
  end_calc_datetime = Time.now.change({ hour: 9, min: 0, sec: 0, usec: 0 }).iso8601

  fourteen_days_ago = 14.days.ago.iso8601

  thirty_five_years_ago = 35.years.ago.iso8601
  thirty_years_ago = 30.years.ago.iso8601
  twenty_five_years_ago = 25.years.ago.iso8601

  user_role_id = user_role_object_id
  guest_level_id = guest_level_object_id
  regular_level_id = regular_level_object_id
  vip_level_id = vip_level_object_id

  total_users = User.find_by_sql("SELECT * FROM users WHERE role_id = '#{user_role_id}'").count
  total_males = User.find_by_sql("SELECT * FROM users WHERE role_id = '#{user_role_id}' AND gender = 'Male'").count
  total_females = User.find_by_sql("SELECT * FROM users WHERE role_id = '#{user_role_id}' AND gender = 'Female'").count
  total_gender_unknown = User.find_by_sql("SELECT * FROM users WHERE role_id = '#{user_role_id}' AND gender IS NULL").count
  total_guests = User.find_by_sql("SELECT * FROM users WHERE role_id = '#{user_role_id}' AND loyaltylevel_id = '#{guest_level_id}'").count
  total_regulars = User.find_by_sql("SELECT * FROM users WHERE role_id = '#{user_role_id}' AND loyaltylevel_id = '#{regular_level_id}'").count
  total_vips = User.find_by_sql("SELECT * FROM users WHERE role_id = '#{user_role_id}' AND loyaltylevel_id = '#{vip_level_id}'").count
  age_35plus = User.find_by_sql("SELECT * FROM users WHERE role_id = '#{user_role_id}' AND dob <= '#{thirty_five_years_ago}'").count
  age_3034 = User.find_by_sql("SELECT * FROM users WHERE role_id = '#{user_role_id}' AND dob BETWEEN '#{thirty_five_years_ago}' AND '#{thirty_years_ago}'").count
  age_2529 = User.find_by_sql("SELECT * FROM users WHERE role_id = '#{user_role_id}' AND dob BETWEEN '#{thirty_years_ago}' AND '#{twenty_five_years_ago}'").count
  age_2124 = User.find_by_sql("SELECT * FROM users WHERE role_id = '#{user_role_id}' AND dob >= '#{twenty_five_years_ago}'").count

  active_users = Timeline.find_by_sql("
    SELECT DISTINCT user_id FROM timelines
    (event_type = 'Credit Earned' OR event_type = 'Reward Redeemed') AND
    date >= '#{fourteen_days_ago}'
  ").count

  traffic = Timeline.find_by_sql("
    SELECT * FROM timelines
    WHERE event_type = 'Credit Earned' AND
    date BETWEEN '#{start_calc_datetime}' AND '#{end_calc_datetime}'
  ").count

  rewards_redeemed = RewardsUsers.find_by_sql("
    SELECT * FROM rewards_users
    WHERE user_has_redeemed = 1 AND
    redeemed_date BETWEEN '#{start_calc_datetime}' AND '#{end_calc_datetime}'
  ").count

  save_stats(
    total_users,
    total_males,
    total_females,
    total_gender_unknown,
    total_guests,
    total_regulars,
    total_vips,
    age_35plus,
    age_3034,
    age_2529,
    age_2124,
    active_users,
    traffic,
    rewards_redeemed
  )
end

puts "Running Drop In stats"
calc_stats()
